examples/propagate-status-code/src/propagate_status_code.rs

```diff
@@ -0,0 +1,271 @@
+use apollo_router_core::{
+    register_plugin, Plugin, RouterRequest, RouterResponse, SubgraphRequest, SubgraphResponse,
+};
+use http::StatusCode;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use tower::{util::BoxService, BoxError, ServiceExt};
+
+#[derive(Serialize, Deserialize, JsonSchema)]
+struct PropagateStatusCodeConfig {
+    status_codes: Vec<u16>,
+}
+
+#[derive(Default)]
+// Global state for our plugin would live here.
+// We don't need any in this example
+struct PropagateStatusCode {
+    status_codes: Vec<u16>,
+}
+
+impl Plugin for PropagateStatusCode {
+    // We either forbid anonymous operations,
+    // Or we don't. This is the reason why we don't need
+    // to deserialize any configuration from a .yml file.
+    //
+    // Config is a unit, and `ForbidAnonymousOperation` derives default.
+    type Config = PropagateStatusCodeConfig;
+
+    fn new(configuration: Self::Config) -> Result<Self, BoxError> {
+        Ok(Self {
+            status_codes: configuration.status_codes,
+        })
+    }
+
+    fn subgraph_service(
+        &mut self,
+        _name: &str,
+        service: BoxService<SubgraphRequest, SubgraphResponse, BoxError>,
+    ) -> BoxService<SubgraphRequest, SubgraphResponse, BoxError> {
+        let all_status_codes = self.status_codes.clone();
+        service
+            .map_response(move |res| {
+                if all_status_codes.contains(&res.response.status().as_u16()) {
+                    res.context
+                        .upsert(
```

Comment:
I'm currently updating the comments before i mark it as ready to review, but the way i saw it was this:

```yml
plugins:
  example.propagate_status_code:
    # Status codes are represented in order,
    # if a subgraph returns a 400
    # and an other one returns a 401
    # the response status code will be 401
    status_codes:
      - 401
      - 403
      - 400
      - 500
```

otherwise we would run into all sorts of races if we fetch things in parallel, and several codes are in our watchlist

Reply:
Ahh, as you have the status codes your config and they were already checked you can just upsert with the status of this request. Even simpler :)

Reply:
I might have misunderstood but how would that solve this test case:

- Parallel fetches to `a` and `b`
- `a` returns a http 403
- `b` returns an http 401

expected status_code => http 401

this is the reason why i upsert with an array, however I could check for indexes and always keep the "most important around" if that's what you meant?